onchain/src/tests/hodl_quest.cairo

```diff
@@ -0,0 +1,74 @@
+use art_peace::{IArtPeaceDispatcher, IArtPeaceDispatcherTrait};
+use art_peace::quests::hodl_quest::HodlQuest::HodlQuestInitParams;
+use art_peace::quests::interfaces::{IQuestDispatcher, IQuestDispatcherTrait};
+
+use art_peace::tests::art_peace::deploy_with_quests_contract;
+use art_peace::tests::utils;
+use starknet::{ContractAddress, get_caller_address, get_contract_address, contract_address_const};
+use snforge_std as snf;
+use snforge_std::{declare, CheatTarget, ContractClassTrait};
+
+const reward_amt: u32 = 18;
+const user_extra_pixel: u32 = 0;
+
+
+fn deploy_hodl_quest() -> ContractAddress {
+    let contract = declare("HodlQuest");
+
+    let mut hodl_quest_calldata = array![];
+    HodlQuestInitParams {
+        art_peace: utils::ART_PEACE_CONTRACT(), reward: reward_amt, extra_pixel: user_extra_pixel,
+    }
+        .serialize(ref hodl_quest_calldata);
+
+    contract.deploy(@hodl_quest_calldata).unwrap()
+}
+
+
+#[test]
+fn deploy_hodl_quest_test() {
+    let hodl_quest = deploy_hodl_quest();
+    let art_peace = IArtPeaceDispatcher {
+        contract_address: deploy_with_quests_contract(array![].span(), array![hodl_quest].span())
+    };
+
+    let zero_address = contract_address_const::<0>();
+
+    assert!(
+        art_peace.get_days_quests(0) == array![zero_address, zero_address, zero_address].span(),
+        "Daily quests were not set correctly"
+    );
+    assert!(
+        art_peace.get_main_quests() == array![hodl_quest].span(),
+        "Main quests were not set correctly"
+    );
+}
+
+
+#[test]
+fn hodl_quest_test() {
+    let hodl_quest_contract_address = deploy_hodl_quest();
+
+    let art_peace = IArtPeaceDispatcher {
+        contract_address: deploy_with_quests_contract(
+            array![].span(), array![hodl_quest_contract_address].span()
+        )
+    };
+
+    let calldata: Array<felt252> = array![];
+    snf::start_prank(CheatTarget::One(art_peace.contract_address), utils::PLAYER1());
+
+    snf::store(
+        hodl_quest_contract_address,
+        selector!("user_extra_pixel"),
```

Comment:
You should be setting the users `extra_pixels`. Here is an example using store to set a value in a map. https://github.com/keep-starknet-strange/blobstream-starknet/blob/842ee5ad46dede70ec8bf5d1615426e44e3e08cc/src/tests/test_blobstreamx.cairo#L62

Reply:
I follow the example above, but I did not understand it, 

Can you provide a sample example on this 
@b-j-roberts

Reply:
Not sure what you mean, but yours should be something like 
```
snf::store(
        art_peace_contract,
        snf::map_entry_address(
            selector!("extra_pixels"), array![users address].span()
        ),
        array![value]
      );
```

Reply:
Ok

Reply:
`self.extra_pixel.read() ` 

does not get updated when i use  

```
snf::store(
        art_peace.contract_address,
        snf::map_entry_address(selector!("extra_pixels"), array![18].span()),
        array![18].span()
    );
```

i go the following error 

```
Failure data:
    0x5175657374206e6f7420636c61696d61626c65 ('Quest not claimable')
```
when i ran scarb test @b-j-roberts

Reply:
The users address should not be 18 but the user you are adding the pixels for

Reply:
```
 snf::store(
        art_peace.contract_address,
        snf::map_entry_address(selector!("extra_pixels"), array![contract_address_const::<1>()].span()),
        array![18].span()
    );
```

the array![contract_address_const::<1>()].span())  return this error 


`Unexpected argument type. Expected: "core::array::Span::<core::felt252>", found: "core::array::Span::<core::starknet::contract_address::ContractAddress>".`

Reply:
You can use into on an address to convert to felt

Reply:
its passed,

Reply:
You need to make sure to set the value to the number of extra pixels to set, then the user address to the player.
You can use `println!` while debugging to try and trace where your error is coming from

Reply:
@b-j-roberts

Reply:
this line  

  ```
snf::store(
        art_peace.contract_address,
        snf::map_entry_address(selector!("extra_pixels"), array![65].span()),
        array![user.into()].span()
    );
```

does not update extra_pixels map

```

 // Map: user's address -> amount of extra pixels they have
        extra_pixels: LegacyMap::<ContractAddress, u32>,
```

Reply:
this line keep returning 0 


```
 let get_extra_pixels_count = art_peace_main.get_user_extra_pixels_count(user);
            println!("get_extra_pixels_count: {} ", get_extra_pixels_count);
```


which mean  this 


```
  snf::store(
        art_peace.contract_address,
        snf::map_entry_address(selector!("extra_pixels"), array![65].span()),
        array![user.into()].span()
    );

```

does not update the user extra_pixel map

```
 // Map: user's address -> amount of extra pixels they have
        extra_pixels: LegacyMap::<ContractAddress, u32>,

```
@TAdev0 can you help here

Reply:
the pixel metadata field

 =>

 `extra_pixels: LegacyMap::<ContractAddress, u32>,` 

is updated by the call of the contract when we deploy the contract, for this to work we need to update it, then 

`get_user_extra_pixels_count`

 can be able to return extra-pixel the user save which will be claimable

Reply:
`claim_main_quest`

 is using 

`starknet::get_caller_address(); `

to update extra_pixels if the user has a reward > 0, 

but our 

`is_claimable`

 is using 

`user: ContractAddress,`

 to get the get_extra_pixels_count from 

`art_peace_main.get_user_extra_pixels_count(user.into());`